tests/utilities/test_cli.py

```diff
@@ -688,10 +712,172 @@ def __init__(self, optim1: dict, optim2: dict, scheduler: dict):
+            self.called.append("after_predict")
+
+        def before_tune(self):
+            self.called.append("before_tune")
+
+        def tune(self, **_):
+            self.called.append("tune")
+
+        def after_tune(self):
+            self.called.append("after_tune")
+
+    with mock.patch("sys.argv", ["any.py", fn]):
+        cli = TestCLI(BoringModel)
+    assert cli.called == [f"before_{fn}", fn, f"after_{fn}"]
+
+
+def test_lightning_cli_subcommands():
+    subcommands = LightningCLI.subcommands()
+    trainer = Trainer()
+    for subcommand, exclude in subcommands.items():
+        fn = getattr(trainer, subcommand)
+        parameters = list(inspect.signature(fn).parameters)
+        for e in exclude:
+            # if this fails, it's because the parameter has been removed from the associated `Trainer` function
+            # and the `LightningCLI` subcommand exclusion list needs to be updated
+            assert e in parameters
+
+
+def test_lightning_cli_custom_subcommand():
+    class TestTrainer(Trainer):
+        def foo(self, model: LightningModule, x: int, y: float = 1.0):
+            """
+            Sample extra function.
+
+            Args:
+                model: A model
+                x: The x
+                y: The y
+            """
+
+    class TestCLI(LightningCLI):
+        @staticmethod
+        def subcommands():
+            subcommands = LightningCLI.subcommands()
+            subcommands["foo"] = {"model"}
+            return subcommands
+
+    out = StringIO()
+    with mock.patch("sys.argv", ["any.py", "-h"]), redirect_stdout(out), pytest.raises(SystemExit):
+        TestCLI(BoringModel, trainer_class=TestTrainer)
+    out = out.getvalue()
+    assert "Sample extra function." in out
+    assert "{fit,validate,test,predict,tune,foo}" in out
+
+    out = StringIO()
+    with mock.patch("sys.argv", ["any.py", "foo", "-h"]), redirect_stdout(out), pytest.raises(SystemExit):
+        TestCLI(BoringModel, trainer_class=TestTrainer)
+    out = out.getvalue()
+    assert "A model" not in out
+    assert "Sample extra function:" in out
+    assert "--x X" in out
+    assert "The x (required, type: int)" in out
+    assert "--y Y" in out
+    assert "The y (type: float, default: 1.0)" in out
+
+
+def test_lightning_cli_run():
+    with mock.patch("sys.argv", ["any.py"]):
+        cli = LightningCLI(BoringModel, run=False)
+    assert cli.trainer.global_step == 0
     assert isinstance(cli.trainer, Trainer)
     assert isinstance(cli.model, LightningModule)
+
+    with mock.patch("sys.argv", ["any.py", "fit"]):
+        cli = LightningCLI(BoringModel, trainer_defaults={"max_steps": 1, "max_epochs": 1})
+    assert cli.trainer.global_step == 1
+    assert isinstance(cli.trainer, Trainer)
+    assert isinstance(cli.model, LightningModule)
+
+
+def test_lightning_cli_config_with_subcommand(tmpdir):
+    config = {
+        # FIXME: this is not necessary, I expected it to be
+        # "subcommand": "test",
```

Comment:
@mauvilsa Why does this implicitly assume the subcommand "test"?
Shouldn't adding `"subcommand": "test"` be required?

Probably linked to my comment below

Reply:
Explicit subcommand not required.